### ISSUE TITLE
Feature: Query and List Rooms in Hubs Admin

### ISSUE THREAD

**Issue by yalegria** [RutgersGRID/hubs#34]

**Comment by Voxelghiest** [comment]
I've created a new branch, `admin-room-listing`, to work on this feature. Here's my current plan for implementing a way to query information about the currently active rooms.

The Hubs Admin webpage is composed of a number of panels, each of which displays information about a different aspect of the Hubs server (scenes, users, avatars, etc.). These are implemented in the `src/react-components` folder of Hubs Admin, and then added to the admin page in `admin.js` as a [Resource](https://marmelab.com/react-admin/Resource.html) component, which is a feature of [React-Admin](https://marmelab.com/react-admin/), a React library for creating admin pages.

To summarize, in order to implement this feature, I plan to:
1. Implement the functionality for the room listing panel as a new component in `src/react-components`
2. Add the new panel to the admin page by modifying `admin.js`

**Comment by Voxelghiest** [comment]
In order to load the data for a Resource component, React-Admin queries the Reticulum server, which communicates with the PostgREST API to access the Postgres database containing the data. To query the currently active rooms from this database, we need to know the name of the table with the room data. I accessed the `ret_dev` database (which manages all the data) and listed all of the tables it has, sorted alphabetically. Here's a snippet of that list:


![Image](https://user-images.githubusercontent.com/72222982/206532905-d6450460-3c93-4bff-9db4-4c4792129917.png)

There isn't anything listed as "rooms". "room-objects" refers to the interactive objects that can be dropped into a scene, so it's not that.

I need to find what the name of the room table is, or find out why Reticulum seemingly doesn't keep track of open rooms in the `ret_dev` database.

**Comment by Voxelghiest** [comment]
I figured it out. In Reticulum, individual rooms are actually called [Hubs](https://github.com/mozilla/reticulum/blob/master/lib/ret/hub.ex). And there is a "hubs" table entry listed in the `ret_dev` database! I printed out the details of the "hubs" table below, so we have a reference for all the fields it contains:

![Image](https://user-images.githubusercontent.com/72222982/206538932-db842715-d437-45b2-bc23-fd08ba4ae6a1.png)

Now that I know which table to query for, I can begin implementing the code for the React-Admin Resource component that will access this table and get a list of all active rooms.

**Comment by Voxelghiest** [comment]
I believe I have finished implementing the new feature. I pushed my code to the `admin-room-listing` branch. To aid in editing the code, I moved the server files into a shared folder in the virtual machine, so before I can test the new code, I have to reinstall the node modules associated with Hubs Client and Hubs Admin.

**Comment by Voxelghiest** [comment]
I've run into an error having to do with symlinks while running `npm install`:

![Image](https://user-images.githubusercontent.com/72222982/209164330-d52f9e6f-9088-45b9-8cb7-209f7831cac2.png)

I spent a long time trying to figure it out, but I think I have the solution. Because the Ubuntu VM is trying to use a symlink stored on a MacOS system (through the VBox shared folder), it doesn't work. Thankfully, you can apparently run `npm install` with the `--no-bin-links` option to prevent it from creating symlinks. I'm going to try that and see if it works. If not, then I'll simply switch back to keeping all the files on the VM itself.

For reference, I found the `--no-bin-links` solution [here](https://stackoverflow.com/questions/39098812/npm-install-when-to-use-no-bin-links).

**Comment by Voxelghiest** [comment]
I ran into some other random issue:

![Image](https://user-images.githubusercontent.com/72222982/209169692-617c7dd4-e8a0-43a5-9349-76651ec2dc93.png)

It's fine. I'll just switch the files back to being located completely on the virtual machine.

**Comment by Voxelghiest** [comment]
Alrighty. Switching back fixed the issues with `npm install`. Now I should be able to test out the new code I wrote.

**Comment by Voxelghiest** [comment]
I set everything up again, started Reticulum, Dialog, Hubs Client, and Hubs Admin, and made a new room (to check if the Rooms list updates appropriately). However, upon opening the new panel on the admin page, I encountered an error:

![Image](https://user-images.githubusercontent.com/72222982/209189195-78bc55ae-ecd9-4fe7-a0c8-08cdb812ca4a.png)

I'm pretty sure the "relation does not exist" error has to do with the database table I chose. I'll see if I can find the correct one.

**Comment by Voxelghiest** [comment]
Okay, so it turns out that "ret0_admin" is a different thing than "ret_dev".

"ret0_admin" is, in PSQL terms, what's called a *schema*. PSQL is organized into a hierarchy of _databases_, which contain a number of _schemas_, which contain a number of _tables_. "ret_dev" is the database containing all the information that Hubs uses, whereas "ret0_admin" is the schema that Hubs Admin is accessing to display its data.

When I was looking through the ret_dev database earlier, I was using the "public" schema, which is the default. However, when I switched to the the ret0_admin schema, these were the available relations:

![Image](https://user-images.githubusercontent.com/72222982/209195324-c827a265-c491-46a6-b62f-a03c963991c0.png)

Notably, these are listed as being *views*, not tables. I still have to figure out what a view is. The other important fact is that "hubs" is not listed as a view, which I guess means that this schema doesn't track the active rooms. There might be a way to set that up, though.

To summarize: My next task is to learn what a "view" is, and then figure out if there is a way to create a view for the currently active rooms.

**Comment by Voxelghiest** [comment]
So a "view" is essentially a way of packaging a custom PSQL query on one or more tables into one reusable unit. A detailed description on how to create and modify views can be found [here](https://www.enterprisedb.com/postgres-tutorials/how-create-modify-delete-and-manage-views-postgresql). For example, the "accounts" view in the ret0_admin schema is defined as follows:

![Image](https://user-images.githubusercontent.com/72222982/209197869-7ba4cd80-7682-4e52-ac14-73f948f1b40f.png)

If I create a new view for the "hubs" table in the ret0_admin schema, I should be able to access the room data.

**Comment by Voxelghiest** [comment]
For future reference, this is the view definition I made:
```sql
CREATE VIEW hubs AS SELECT 
    o.hub_id AS id, o.hub_sid, o.name, o.inserted_at, o.updated_at, o.max_occupant_count, 
    o.entry_mode, o.spawned_object_types, o.scene_id, o.host, o.created_by_account_id, 
    o.scene_listing_id, o.last_active_at, o.embed_token, o.embedded, o.member_permissions, 
    o.allow_promotion, o.description, o.room_size
FROM ret0.hubs o;
```

**Comment by Voxelghiest** [comment]
After adding the above view to the ret0_admin schema, I retried accessing the rooms panel on the admin page, but it ended up kicking me out and requesting for me to sign in again, even though I had previously signed in. This persisted even after logging out and logging back in, and it only seemed to affect the Rooms panel.

However! Looking at the Reticulum console, I noticed that the PostgREST request to the database was being made, and returned successfully! Here's the full Reticulum output for when I accessed the Rooms panel (the second-to-last query is the one for the rooms data):

![Image](https://user-images.githubusercontent.com/72222982/209206351-060ef782-2a2f-43b5-a3bb-ed784ac3a383.png)

So that's a step in the right direction. I'm unsure why the rooms panel (and only the rooms panel) seems to cause some sort of authorization error that leads to me getting kicked out of the admin page. That is my next area of investigation.

**Comment by Voxelghiest** [comment]
I think I've figured it out. I looked into the browser console and saw these messages:

![Image](https://user-images.githubusercontent.com/72222982/209214886-ab5669f7-54af-4714-8516-6badc6d0b65f.png)

After some research, I was able to go back into the PSQL database and check the permissions for the new view I made in the ret0_admin schema:

![Image](https://user-images.githubusercontent.com/72222982/209215070-36f3ce59-3d1a-4ed5-b30a-4f7278659aa0.png)

As can be seen, the "hubs" view doesn't have the same access privileges as the other views. I'll see if I can correct that and get everything working.

**Comment by Voxelghiest** [comment]
I was able to grant privileges to the ret_admin user with the following command in the PSQL shell (which I found [here](https://www.postgresql.org/docs/13/ddl-priv.html)):
```sql
GRANT SELECT,UPDATE,INSERT,DELETE ON hubs TO ret_admin;
```
Hopefully that works.

**Comment by Voxelghiest** [comment]
Success! Everything went through, and I was able to see a list of all the rooms that have been made so far in our Hubs server:

![Image](https://user-images.githubusercontent.com/72222982/209218679-c31e1abc-c152-4194-a971-71745a73e77d.png)

Now all that's left to do is to modify what information is shown in the rooms panel to whatever we want, such as [how many people are currently in each room](#35).
